test/system/index.js

```diff
@@ -1,5 +1,60 @@
 const trader = require('./trader');
 const config = require('./config.json');
-new trader(config);
+const traders = config.traders;
+const async = require('async');
 
+function createTraders(traders) {
+  const list = [];
+  for (let t of traders) {
+    list.push(new trader(t));
+  }
+  return Promise.resolve(list);
+}
+
+function runSequence1(trader1, trader2) {
+  async.series(
```

Comment:
Use of async really complicates things. This series can use the promises returned like this:
```javascript
trader1.init_state(trader1.config)
  .then(() => trader2.init_state(trader2.config))
  .then(runScenario1)
  .then(runScenario2)
  .then(() => trader1.cancelAllOrders())
  .then(() => trader2.cancelAllOrders())
```

The scenario functions themselves will be made up of Promise chains...
```javascript
function runScenario1() {
  return trader1.placeSomeOrders()
    .then(() => trader2.nextThing())
    .then(() => trader2.andSoOn())
    // ...
}
```

Reply:
Since, we are executing 5 `then` functions after `trader1.init_state` is resolved but we still cannot guarantee that the operation for each `then` will be completed in sequence. For example: `runScenario1` and `runScenario2` both are asynchronous operations and we cannot always ensure that `runScenario1` will be completed before `runScenario2` although the way we have specified the `then`  function will execute `runScenario1` first, and then `runScenario2`.

Reply:
yes, we can. That's how promises work. runScenario1, runScenario2 and any others will each return a promise that resolves when they're complete. the .then() calls will wait for runScenario1 to resolve before it calls runScenario2

Reply:
Ok do you want me to follow this pattern 

` trader1.init_state(trader1.config)
    .then(() => trader2.init_state(trader2.config)
      .then(runScenario1(trader1, trader2)
        .then(() => trader1.subscribeToMessages()
          .then(() => trader2.subscribeToMessages()
            .then(runScenario2(trader1, trader2)
              .then(() => trader1.cancelAllOrders()
                .then(() => trader2.cancelAllOrders()
                  .then(() => trader1.init_state(trader1.config)
                    .then(() => trader2.init_state(trader2.config))
                  )
                )
              )
            )
          )
        )
      )
    );`

Reply:
No - that's a Promise version of the "Pyramid of Doom".
If you assume that runScenario1, runScenario2, ... runScenarioN have access to trader1 and trader2, then the code can look like this:
```javascript
initAllTraders()
  .then(runScenario1) // notice, not calling it here, just passing the function to .then
  .then(runScenario2)
  // ...
  .then(runScenarioN)
  .then(cleanUp)
  .catch(error => {
    console.error(error.stack);
  });
```

A more structured approach means we would pass the traders to the scenarios so we can have each scenario in its own js file. So the scenario functions would return a .then handler function:
```javascript
function runScenario1(trader1) {
  // returns a .then handler
  return function() {
    // here we return a Promise that resolves when this scenario is complete
    return new Promise((resolve, reject) => {
      // do things with trader1 then resolve or reject
    });
  }
}
```

Then the main sequence would look like this...

```javascript
initAllTraders()
  .then(runScenario1(trader1)) // runScenario1 returns a function that will be called by .then
  .then(runScenario2(trader1, trader2))
  // ...
  .then(runScenarioN(trader1, trader2, trader3))
  .then(cleanUp)
  .catch(error => {
    console.error(error.stack);
  });
```